VoodooPS2Trackpad/VoodooPS2Elan.cpp

```diff
@@ -0,0 +1,2220 @@
+            setTouchPadEnable(true);
+            break;
+    }
+}
+
+void ApplePS2Elan::registerHIDPointerNotifications() {
+    IOServiceMatchingNotificationHandler notificationHandler = OSMemberFunctionCast(IOServiceMatchingNotificationHandler, this, &ApplePS2Elan::notificationHIDAttachedHandler);
+
+    // Determine if we should listen for USB mouse attach events as per configuration
+    if (_processusbmouse) {
+        // USB mouse HID description as per USB spec: http://www.usb.org/developers/hidpage/HID1_11.pdf
+        OSDictionary *matchingDictionary = serviceMatching("IOUSBInterface");
+
+        propertyMatching(OSSymbol::withCString(kUSBHostMatchingPropertyInterfaceClass), OSNumber::withNumber(kUSBHIDInterfaceClass, 8), matchingDictionary);
+        propertyMatching(OSSymbol::withCString(kUSBHostMatchingPropertyInterfaceSubClass), OSNumber::withNumber(kUSBHIDBootInterfaceSubClass, 8), matchingDictionary);
+        propertyMatching(OSSymbol::withCString(kUSBHostMatchingPropertyInterfaceProtocol), OSNumber::withNumber(kHIDMouseInterfaceProtocol, 8), matchingDictionary);
+
+        // Register for future services
+        usb_hid_publish_notify = addMatchingNotification(gIOFirstPublishNotification, matchingDictionary, notificationHandler, this, NULL, 10000);
+        usb_hid_terminate_notify = addMatchingNotification(gIOTerminatedNotification, matchingDictionary, notificationHandler, this, NULL, 10000);
+        OSSafeReleaseNULL(matchingDictionary);
+    }
+
+    // Determine if we should listen for bluetooth mouse attach events as per configuration
+    if (_processbluetoothmouse) {
+        // Bluetooth HID devices
+        OSDictionary *matchingDictionary = serviceMatching("IOBluetoothHIDDriver");
+        propertyMatching(OSSymbol::withCString(kIOHIDVirtualHIDevice), kOSBooleanFalse, matchingDictionary);
+
+        // Register for future services
+        bluetooth_hid_publish_notify = addMatchingNotification(gIOFirstPublishNotification, matchingDictionary, notificationHandler, this, NULL, 10000);
+        bluetooth_hid_terminate_notify = addMatchingNotification(gIOTerminatedNotification, matchingDictionary, notificationHandler, this, NULL, 10000);
+        OSSafeReleaseNULL(matchingDictionary);
+    }
+}
+
+void ApplePS2Elan::unregisterHIDPointerNotifications() {
+    // Free device matching notifiers
+    // remove() releases them
+
+    if (usb_hid_publish_notify) {
+        usb_hid_publish_notify->remove();
+    }
+
+    if (usb_hid_terminate_notify) {
+        usb_hid_terminate_notify->remove();
+    }
+
+    if (bluetooth_hid_publish_notify) {
+        bluetooth_hid_publish_notify->remove();
+    }
+
+    if (bluetooth_hid_terminate_notify) {
+        bluetooth_hid_terminate_notify->remove();
+    }
+
+    attachedHIDPointerDevices->flushCollection();
+}
+
+void ApplePS2Elan::notificationHIDAttachedHandlerGated(IOService *newService, IONotifier *notifier) {
+    char path[256];
+    int len = 255;
+    memset(path, 0, len);
```

Comment:
Why is the last byte not zeroed?

Reply:
I didn't write it. It was copied from `ApplePS2SynapticsTouchPad::notificationHIDAttachedHandlerGated`

Reply:
https://developer.apple.com/documentation/kernel/ioregistryentry/1811074-getpath

It looks like there is no need for zeroing the buffer and since the docs doesn't say anthing else, I believe we can assume this is normal c-string, therefore length passed to GetPath can be total buffer length (256 in this case) and assume the method will put zero byte at the end of the buffer.